airbyte-config/models/src/main/java/io/airbyte/config/storage/CloudStorageConfigs.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) 2021 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.config.storage;
+
+import com.google.common.base.Preconditions;
+
+/**
+ * Immutable configuration for configuring Cloud storage clients. We usually allow configuring one
+ * of 3 type of cloud storage clients in our env variables. We then want to opaquely pass that to
+ * wherever that cloud storage is used and then, based on the configuration, spin up the correct
+ * client. This configuration object allows us to do that.
+ */
+public class CloudStorageConfigs {
+
+  public enum WorkerStorageType {
+    S3,
+    MINIO,
+    GCS
+  }
+
+  private final WorkerStorageType type;
+  private final S3Config s3Config;
+  private final MinioConfig minioConfig;
+  private final GcsConfig gcsConfig;
+
+  public static CloudStorageConfigs s3(final S3Config config) {
+    return new CloudStorageConfigs(WorkerStorageType.S3, config, null, null);
+  }
+
+  public static CloudStorageConfigs minio(final MinioConfig config) {
+    return new CloudStorageConfigs(WorkerStorageType.MINIO, null, config, null);
+  }
+
+  public static CloudStorageConfigs gcs(final GcsConfig config) {
+    return new CloudStorageConfigs(WorkerStorageType.GCS, null, null, config);
+  }
+
+  private CloudStorageConfigs(final WorkerStorageType type,
+                              final S3Config s3Config,
+                              final MinioConfig minioConfig,
+                              final GcsConfig gcsConfig) {
+    validate(type, s3Config, minioConfig, gcsConfig);
+
+    this.type = type;
+    this.s3Config = s3Config;
+    this.minioConfig = minioConfig;
+    this.gcsConfig = gcsConfig;
+  }
+
+  private void validate(final WorkerStorageType type,
+                        final S3Config s3Config,
+                        final MinioConfig minioConfig,
+                        final GcsConfig gcsConfig) {
+    switch (type) {
+      case S3 -> {
+        Preconditions.checkNotNull(s3Config);
+        Preconditions.checkArgument(minioConfig == null);
+        Preconditions.checkArgument(gcsConfig == null);
+      }
+      case MINIO -> {
+        Preconditions.checkArgument(s3Config == null);
+        Preconditions.checkNotNull(minioConfig);
+        Preconditions.checkArgument(gcsConfig == null);
+      }
+      case GCS -> {
+        Preconditions.checkArgument(s3Config == null);
+        Preconditions.checkArgument(minioConfig == null);
+        Preconditions.checkNotNull(gcsConfig);
+      }
+    }
+  }
+
+  public WorkerStorageType getType() {
+    return type;
+  }
+
+  public S3Config getS3Config() {
+    return s3Config;
+  }
+
+  public MinioConfig getMinioConfig() {
+    return minioConfig;
+  }
+
+  public GcsConfig getGcsConfig() {
+    return gcsConfig;
+  }
+
+  public static abstract class S3LikeWorkerStorageConfig {
```

Comment:
garbage name.

Reply:
`S3CompatibleWorkerStorageConfig`? Also it could just be a `S3WorkerStorageConfig` or `S3ApiWorkerStorageConfig` because S3 is the API, not the technology.

Reply:
I like `S3ApiWorkerStorageConfig`.